packages/pipeline/src/parsers/copper/index.ts

```diff
@@ -0,0 +1,274 @@
+ *
+ * @param opportunities - The array of opportunities returned from the API
+ * @returns Returns an array of Copper Opportunity entities
+ */
+export function parseOpportunities(opportunities: CopperOpportunityResponse[]): CopperOpportunity[] {
+    return opportunities.map(opp => {
+        const entity = new CopperOpportunity();
+        entity.id = opp.id;
+        entity.name = opp.name;
+        entity.assigneeId = opp.assignee_id;
+        entity.closeDate = opp.close_date;
+        entity.companyId = opp.company_id;
+        entity.companyName = opp.company_name;
+        entity.customerSourceId = opp.customer_source_id;
+        entity.lossReasonId = opp.loss_reason_id;
+        entity.pipelineId = opp.pipeline_id;
+        entity.pipelineStageId = opp.pipeline_stage_id;
+        entity.primaryContactId = opp.primary_contact_id;
+        entity.priority = opp.priority;
+        entity.status = opp.status;
+        entity.tags = opp.tags;
+        entity.interactionCount = opp.interaction_count;
+        entity.monetaryValue = opp.monetary_value;
+        entity.winProbability = opp.win_probability;
+        entity.dateCreated = opp.date_created * ONE_SECOND;
+        entity.dateModified = opp.date_modified * ONE_SECOND;
+        entity.customFields = opp.custom_fields.map(f => f.custom_field_definition_id);
+        return entity;
+    });
+}
+
+/**
+ * Parse response from Copper API /activity_types/
+ *
+ * @param activityTypeResponse - Activity Types response from the API, keyed by "user" or "system"
+ * @returns Returns an array of Copper Activity Type entities
+ */
+export function parseActivityTypes(
+    activityTypeResponse: Map<CopperActivityTypeCategory, CopperActivityTypeResponse[]>,
+): CopperActivityType[] {
+    const values: CopperActivityTypeResponse[] = R.reduce(
+        (acc, val) => acc.concat(val),
+        [] as CopperActivityTypeResponse[],
+        Object.values(activityTypeResponse),
+    );
+    return values.map(activityType => {
+        const parsed: CopperActivityType = {
+            id: activityType.id,
+            name: activityType.name,
+            category: activityType.category.toString(),
+            isDisabled: activityType.is_disabled,
+            countAsInteraction: activityType.count_as_interaction,
+        };
+        return parsed;
+    });
+}
+
+/**
+ * Parse response from Copper API /custom_field_definitions/
+ *
+ * @param customFieldResponse - array of custom field definitions returned from the API"
+ * @returns Returns an array of Copper Custom Field entities
+ */
+export function parseCustomFields(customFieldResponse: CopperCustomFieldResponse[]): CopperCustomField[] {
+    function parseTopLevelField(field: CopperCustomFieldResponse): CopperCustomField[] {
+        const parsed: CopperCustomField = {
+            id: field.id,
+            name: field.name,
+            dataType: field.data_type.toString(),
+        };
+
+        if (field.options !== undefined) {
```

Comment:
I would appreciate a comment about what's going on here. It seems pretty weird to pull some properties out of `field` and then concatenate them with in a list which basically already contains `field` itself. Could you clarify?

Reply:
thanks, on second thought this *is* confusing. the API response consists of a top level field and (potentially) some nested fields. I added a comment and simplified the code somewhat.